### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by rishabhhj**

### CODE REVIEW

File: testng.xml [hunk 1]
```diff
@@ -5,24 +5,20 @@
 	<test name="TestHeadless" thread-count="2" parallel="classes">
```
Comment (4shivanshu) [inline testng.xml:5]: Thread Count: The thread-count="2" is fine for parallel execution but consider increasing it for large-scale tests if infrastructure supports it.
Parallelization: Classes are being run in parallel. Ensure no shared resources are accessed between tests to avoid flaky results.

File: src/main/java/org/automation/utilities/WebdriverWaits.java [hunk 2]
```diff
@@ -124,14 +122,66 @@ public static void fluentWait_ElementLocated(long waitTimeForTimeout, long waitT
         wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
 
     }
-    
+
     public static void fluentWait_ElementIntactable(long waitTimeForTimeout, long waitTimeForPolling, By locator) {
```
Comment (4shivanshu) [inline src/main/java/org/automation/utilities/WebdriverWaits.java:126]: Improvement Suggestion: Handle TimeoutException and log errors if the element is not interactable.
Good Use of Fluent Wait: Provides flexibility for dynamic wait intervals.

File: src/main/java/org/automation/listeners/TestRunListener.java [hunk 3]
```diff
@@ -66,6 +75,22 @@ public void onStart(ISuite suite) {
...
+
+    private int getPassPercentage(ITestContext context) {
+        int totalTests = context.getAllTestMethods().length;
+        int passedTests = context.getPassedTests().size();
+        return (passedTests * 100) / totalTests;
     }
 
+    @Override
+    public boolean retry(ITestResult result) {
+        if (retryCount < maxRetryCount) {
```
Comment (4shivanshu) [inline src/main/java/org/automation/listeners/TestRunListener.java:89]: Improvement Suggestion: Include a mechanism to handle known intermittent failures based on specific exceptions if possible.

File: src/test/java/scenarios/CustomersTest.java [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+//        WebdriverWaits.sleep(2000);
+//        session.getCustomersPage().setCustomerName("Rishabh");
+//        session.getCustomersPage().doneBtn().click();
+
+        session.getCustomersPage().findAddCustomer().click();
+        WebDriverWait waitt = new WebDriverWait(getDriver(), Duration.ofSeconds(10));
+        waitt.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//span[@class='me-1']")));
+        session.getCustomersPage().getPhoneField().setText("90110");
+        session.getCustomersPage().goPhoneNumber().click();
+        Assertions.assertTrue(session.getCustomersPage().getPhoneValidation().isDisplayed());
```
Comment (4shivanshu) [inline src/test/java/scenarios/CustomersTest.java:54]: Boundary Values: Ensure edge cases such as invalid phone numbers and empty inputs are tested.
Improvement Suggestion: Parameterize test cases using @DataProvider to handle multiple inputs efficiently.

File: src/main/java/org/automation/utilities/WebdriverWaits.java [hunk 5]
```diff
@@ -124,14 +122,66 @@ public static void fluentWait_ElementLocated(long waitTimeForTimeout, long waitT
...
+
+            if (!currentValue.equals(oldValue)) {
+                return ele;
+            } else {
+                return null;
+            }
+        });
     }
+
+    public static void retryClick(WebElement element, int maxRetries) {
```
Comment (4shivanshu) [inline src/main/java/org/automation/utilities/WebdriverWaits.java:164]: Improvement Suggestion: Add a delay between retries to allow the element to become interactable.

File: src/test/java/scenarios/TransactionTest.java [hunk 6]
```diff
@@ -661,29 +691,32 @@ public void a2verifyThatTransactionListAppears() {
...
 
+
         // Check  setAfterFilter contains all elements from setBeforeFilter
         Assertions.assertTrue(setAfterFilter.containsAll(setBeforeFilter));
 
         session.getTransactionsPage().getFilterIcon().click();
         // Click on Download button
 
-        WebdriverWaits.sleep(5000);
         String fileStatus = ActionEngine.isFileDownloaded("Transactions.xlsx");
```
Comment (4shivanshu) [inline src/test/java/scenarios/TransactionTest.java:701]: Improvement Suggestion: Validate file content after download to ensure the file is correct.

File: src/main/java/org/automation/pages/CustomersPage.java [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+    public Clickable getCustPopupCloseBtn() {
+        return Clickable.getElementBy(customerPopupClose,"Close button for Customer popup");
+    }
+
+    public Clickable filter() {
+        return Clickable.getElementBy(filter,"Filter icon");
+    }
+
+    public void setPhoneNumber(String phoneNumberInput) {
+        Editable phoneField = Editable.getElementBy(phoneNumber, "Phone Number field");
```
Comment (4shivanshu) [inline src/main/java/org/automation/pages/CustomersPage.java:108]: Good Practice: Abstracting frequently used actions like setting phone numbers or email addresses into reusable methods.
Improvement Suggestion: Combine actions for better readability, e.g., setAndSubmitPhoneNumber.

File: src/test/java/scenarios/CustomersTest.java (hunk 4 above)
Comment (rishabhhj) [reply src/test/java/scenarios/CustomersTest.java:56]: Thank you for the suggestion! I've updated the code to use the @DataProvider annotation for parameterizing the test cases. This should now handle multiple inputs more efficiently.

File: src/main/java/org/automation/pages/CustomersPage.java (hunk 7 above)
Comment (rishabhhj) [reply src/main/java/org/automation/pages/CustomersPage.java:115]: Thank you so much for your feedback!
I did not integrate the action into a single step to ensure that the input can be modified directly from the test case page at any time.
